reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/WindowSized.kt

```diff
@@ -0,0 +1,96 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.base.setCancellable
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.disposable.DisposableWrapper
+import com.badoo.reaktive.subject.unicast.UnicastSubject
+import com.badoo.reaktive.utils.atomic.AtomicBoolean
+import com.badoo.reaktive.utils.atomic.AtomicInt
+import com.badoo.reaktive.utils.atomic.AtomicLong
+import com.badoo.reaktive.utils.queue.SharedQueue
+
+/**
+ * Please refer to the corresponding RxJava
+ * [document](http://reactivex.io/RxJava/javadoc/io/reactivex/Observable.html#window-long-long-).
+ */
+fun <T> Observable<T>.window(
+    count: Long,
+    skip: Long = count
+): Observable<Observable<T>> {
+    require(count > 0) { "count > 0 required but it was $count" }
+    require(skip > 0) { "skip > 0 required but it was $skip" }
+
+    return observable { emitter ->
+        val upstreamDisposableWrapper = DisposableWrapper()
+
+        val windows = SharedQueue<UnicastSubject<T>>()
+        val activeWindowsCount = AtomicInt(1)
+
+        emitter.setCancellable {
+            if (activeWindowsCount.addAndGet(-1) == 0) {
+                upstreamDisposableWrapper.dispose()
+            }
+        }
+
+        val onTerminate: () -> Unit = {
+            if (activeWindowsCount.addAndGet(-1) == 0 && emitter.isDisposed) {
+                upstreamDisposableWrapper.dispose()
+            }
+        }
+
+        subscribe(object : ObservableObserver<T> {
+
+            private val skippedCount = AtomicLong()
+            private val tailWindowValuesCount = AtomicLong()
```

Comment:
I think just one counter should be enough. Suppose this `counter` would count total amount of emitter items. Then the following math should be valid?
1. We should open a new window if `counter % skip == 0`
2. We should close an oldest window if:
```
val openIndex = counter - count + 1
(openIndex >= 0) && (openIndex % skip == 0)
```

Please correct me if I'm wrong.

Reply:
This math looks correct. Nice!

I spent a lot of time to find a way to use one counter. I checked the [Reactor](https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/FluxWindow.java#L459) and the [RxJava](https://github.com/ReactiveX/RxJava/blob/1ff3cd62091db049b39a2eda0443240e75d6c321/src/main/java/io/reactivex/rxjava3/internal/operators/observable/ObservableWindow.java#L149) implementations and they both use two counters.

Reply:
Sorry for long delay. Could you please provide an example of an overflow?

Reply:
Got it. This is a very good catch, let's use two counters. Sorry for confusing.